californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseStatus.java

```diff
@@ -20,89 +20,78 @@
 package org.eclipse.californium.core.network.stack;
 
 import java.nio.ByteBuffer;
+import java.util.concurrent.ScheduledFuture;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
+import org.eclipse.californium.core.coap.BlockOption;
 import org.eclipse.californium.core.coap.Message;
+import org.eclipse.californium.core.coap.OptionSet;
 
 /**
- * This class represents the status of a blockwise transfer of a request or a
- * response.
- * 
- * This class is package private. Instances of this class are stored inside an
- * exchange and only accessed/modified by the class BlockwiseLayer.
+ * A tracker for the status of a blockwise transfer of a request or response body.
+ * <p>
+ * Instances of this class are accessed/modified by the {@code BlockwiseLayer} only.
  */
-public class BlockwiseStatus {
-
-	public static final int NO_OBSERVE = -1;
+abstract class BlockwiseStatus {
 
-	/** The first token to manage blockwise Observe */
-	private Message first;
-
-	/** The current num. */
-	private int currentNum;
-
-	/** The current szx. */
-	private int currentSzx;
-
-	private boolean randomAccess;
+	private static final Logger LOGGER = Logger.getLogger(BlockwiseStatus.class.getName());
+	private static final int NO_OBSERVE = -1;
 
 	private final int contentFormat;
 
-	/** Indicates whether the blockwise transfer has completed. */
-	private boolean complete;
-
-	private int blockCount;
-
+	protected boolean randomAccess;
 	/*
 	 * It would be nice if we could get rid of this. Currently, the Cf client
 	 * needs it to mark a blockwise transferred notification as such. The
 	 * problem is, that the server includes the observe option only in the first
 	 * block of the notification and we still need to remember it, when the
 	 * last block arrives (block-14).
 	 */
-	/** The observe sequence number of this blockwise transfer */
-	private int observe = NO_OBSERVE;
+	protected int observe = NO_OBSERVE;
+	protected ByteBuffer buf;
```

Comment:
`buf` is accessed sometimes after null check (getBufferSize) and sometimes without (addBlock and getBody).
Though a constructor without createing a buf exists, I think adding IllegalStateException would be a improvement.